### PR TITLE
Automated Keyboard Input

### PR DESCRIPTION
**Pull request by kereliuk**
This provides the testdriver function for sending keyboard input to an element.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/10113)

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build ERRORED

Started: 2018-03-20 18:35:20
Finished: 2018-03-20 19:24:47


<h2>Failing Jobs</h2>
<ul>

<li>lint</li>

<li>firefox:nightly</li>

<li>resources_unittest in py27</li>

<li>chrome:dev</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/26241)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/355997570)

**Comment by jgraham** [comment]
So implementation-wise this looks OK. We need an integration test under `infrastructure` for the new feature. 

In general I'm a little worried about relying on the DWIM methods of WebDriver vs the low level actions.

---

Reviewed 5 of 9 files at r1, 1 of 1 files at r2.
Review status: 5 of 7 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---

*[resources/testdriver.js, line 90 at r2](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113#-L8cCsZ8BfeS2g2ajXYH:-L8cCsZ8BfeS2g2ajXYI:bvi76mp) ([raw file](https://github.com/w3c/web-platform-tests/blob/7578e333cabf4c2ddeaed1a9fc57cbc18502bc02/resources/testdriver.js#L90)):*
> ```JavaScript
>         send_keys: function(element, keys) {
>             if (window.top !== window) {
>                 return Promise.reject(new Error("can only click in top-level window"));
> ```

This isn't a click.

---

*[resources/testdriver.js, line 97 at r2](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113#-L8cD-aO9wpDy3D1N9DT:-L8cD-aQ1iXSnW7ancAH:b-ceaxje) ([raw file](https://github.com/w3c/web-platform-tests/blob/7578e333cabf4c2ddeaed1a9fc57cbc18502bc02/resources/testdriver.js#L97)):*
> ```JavaScript
>             }
> 
>             if (!inView(element)) {
> ```

Doesn't webdriver do this for us?

---

*[tools/wptrunner/wptrunner/executors/protocol.py, line 270 at r2](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113#-L8cCSIt7tTpWpna6N-r:-L8cCSIvDzhgWGr85nii:bxe0rf5) ([raw file](https://github.com/w3c/web-platform-tests/blob/7578e333cabf4c2ddeaed1a9fc57cbc18502bc02/tools/wptrunner/wptrunner/executors/protocol.py#L270)):*
> ```Python
>     @abstractmethod
>     def send_keys(self, element, keys):
>         """Perform a trusted click somewhere on a specific element.
> ```

Comment is wrong.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113#-:-L8cD7oM8uT1iOMgGvxW:b895fw7)*

**Comment by jgraham** [comment]
Reviewed 2 of 9 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113)*

**Comment by gsnedders** [comment]
*[resources/testdriver.js, line 97 at r2](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113#-L8cD-aO9wpDy3D1N9DT:-L8i7FKhEsxLyJ_vVMCQ:bljse4m) ([raw file](https://github.com/w3c/web-platform-tests/blob/7578e333cabf4c2ddeaed1a9fc57cbc18502bc02/resources/testdriver.js#L97)):*
<details><summary><i>Previously, jgraham wrote…</i></summary><blockquote>

Doesn't webdriver do this for us?
</blockquote></details>

As with click, we should do this ourselves, because it means a vendor implementations that is just `window.internals.send_keys(element, keys)` gets the same behaviour as WebDriver

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/w3c/web-platform-tests/10113)*

File: infrastructure/testdriver/send_keys.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+<script src="/resources/testdriver-vendor.js"></script>
+
+<input type="text" id="text">Button</button>
+
+<script>
+async_test(t => {
+  let text_box = document.getElementById("text");
+  test_driver
+    .send_keys(text_box, "Hello, wpt!")
+    .then(() => t.done())
```
Comment (jgraham) [inline infrastructure/testdriver/send_keys.html]: Can we check this actually worked here?

**Review by jgraham — CHANGES_REQUESTED** [review]

File: tools/wptrunner/wptrunner/executors/executormarionette.py [hunk 2]
```diff
@@ -230,7 +231,7 @@ def set(self, name, value):
                 case prefInterface.PREF_BOOL:
                     prefInterface.setBoolPref(pref, value);
                     break;
-                case prefInterface.PREF_INT:
+                case prefInterface.PREF_INT:SendKeysProtocolPart
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/executors/executormarionette.py]: Looks like some junk here.

**Review by jgraham — APPROVED** [review]

**Comment by foolip** [comment]
Travis failures are because of https://github.com/w3c/web-platform-tests/issues/7660

File: infrastructure/testdriver/send_keys.html (hunk 1 above)
Comment (gsnedders) [reply infrastructure/testdriver/send_keys.html]: This still seems like a reasonable comment, @kereliuk!

Comment (gsnedders) [reply infrastructure/testdriver/send_keys.html]: No, GitHub is just showing me random diffs that aren't of the right commit when I first click here. o_O

**Pull request merged**